packages/itwin/tree-widget/src/components/trees/stateless/imodel-content-tree/internal/IModelContentTreeIdsCache.ts

```diff
@@ -0,0 +1,234 @@
+      yield { id: row.id, parentId: row.parentId, targetPartitionId: row.targetPartitionId, hideInHierarchy: !!row.hideInHierarchy };
+    }
+  }
+
+  private async *queryModels(): AsyncIterableIterator<{ id: Id64String; parentId: Id64String }> {
+    const modelsQuery = `
+      SELECT p.ECInstanceId id, p.Parent.Id parentId
+      FROM bis.InformationPartitionElement p
+      INNER JOIN bis.Model m ON m.ModeledElement.Id = p.ECInstanceId
+      WHERE NOT m.IsPrivate
+    `;
+    for await (const row of this._queryExecutor.createQueryReader({ ecsql: modelsQuery }, { rowFormat: "ECSqlPropertyNames", limit: "unbounded" })) {
+      yield { id: row.id, parentId: row.parentId };
+    }
+  }
+
+  private async getSubjectInfos() {
+    this._subjectInfos ??= (async () => {
+      const [subjectInfos, targetPartitionSubjects] = await Promise.all([
+        (async () => {
+          const result = new Map<Id64String, SubjectInfo>();
+          for await (const subject of this.querySubjects()) {
+            const subjectInfo: SubjectInfo = {
+              parentSubject: subject.parentId,
+              hideInHierarchy: subject.hideInHierarchy,
+              childSubjects: new Set(),
+              childModels: new Set(),
+            };
+            if (subject.targetPartitionId) {
+              subjectInfo.childModels.add(subject.targetPartitionId);
+            }
+            result.set(subject.id, subjectInfo);
+          }
+          return result;
+        })(),
+        (async () => {
+          const result = new Map<Id64String, Set<Id64String>>();
+          for await (const model of this.queryModels()) {
+            pushToMap(result, model.id, model.parentId);
+          }
+          return result;
+        })(),
+      ]);
+
+      for (const [subjectId, { parentSubject: parentSubjectId }] of subjectInfos.entries()) {
+        if (parentSubjectId) {
+          const parentSubjectInfo = subjectInfos.get(parentSubjectId);
+          assert(!!parentSubjectInfo);
+          parentSubjectInfo.childSubjects.add(subjectId);
+        }
+      }
+
+      for (const [partitionId, subjectIds] of targetPartitionSubjects) {
+        subjectIds.forEach((subjectId) => {
+          const subjectInfo = subjectInfos.get(subjectId);
+          assert(!!subjectInfo);
+          subjectInfo.childModels.add(partitionId);
+        });
+      }
+
+      return subjectInfos;
+    })();
+    return this._subjectInfos;
+  }
+
+  /** Returns ECInstanceIDs of Subjects that either have direct Model or at least one child Subject with a Model. */
+  public async getParentSubjectIds(): Promise<Id64String[]> {
+    this._parentSubjectIds ??= (async () => {
+      const subjectInfos = await this.getSubjectInfos();
+      const parentSubjectIds = new Set<Id64String>();
+      subjectInfos.forEach((subjectInfo, subjectId) => {
+        if (subjectInfo.childModels.size === 0 && subjectInfo.childSubjects.size === 0) {
```

Comment:
You inverted the condition and added the extra check `childSubjects.size === 0`, should the same be done in `ModelsTreeIdsCache`?

Reply:
> should the same be done in `ModelsTreeIdsCache`?

No. 

Models tree shows Subject nodes only if they have a non-empty Model down in the hierarchy. We use this function to tell if a Subject node has children, so we have to check child modes.

IModel content tree, on the other hand, always shows Subject nodes, so to tell if a Subject has children we check if they have child Subjects or child Models.